src/runtime/device_buffer_utils.h

```diff
@@ -89,7 +89,7 @@ WEAK device_copy make_buffer_copy(const halide_buffer_t *src, bool src_host,
     // Offset the src base pointer to the right point in its buffer.
     c.src_begin = 0;
     for (int i = 0; i < src->dimensions; i++) {
-        c.src_begin += src->dim[i].stride * (dst->dim[i].min - src->dim[i].min);
+        c.src_begin += (uint64_t)src->dim[i].stride * (dst->dim[i].min - src->dim[i].min);
```

Comment:
Just noticed this on the last skim: should this be int64_t ?

Reply:
The fields in `struct device_copy` are `uint64_t` type, including `src_begin`.  Like I said - there's a lack of consistency in the codebase.  This assumes that `stride >= 0` and `dst->dim[i].min >= src->dim[i].min`.  If these assumptions don't hold, then there is something more seriously wrong here I think.

Reply:
I was thinking that the RHS here could be signed (as opposed to the two examples below, where it seems harmless to cast to uint64). But I agree that there are other issues here.

Reply:
This should indeed be int64_t. The operations required to compute an offset given some strides, etc are multiplications, subtractions, additions. None of those differ between signed and unsigned (yay modulo arithmetic), so none of that matters. But there are also upcasts, and those are the only thing that differs between sign and unsigned. The source field is an int32_t, so it must be sign-extended. This PR casts an int32_t to a uint64_t. I'm honestly not sure whether that sign or zero extends. Using an int64_t would make it clear that it's sign-extending it. The rest of the arithmetic is invariant to signed vs unsigned.

There may be bugs elsewhere when the strides are negative, but let's not add a new one.

Reply:
Is `stride` *ever* allowed to be a negative value?  If not, then the PR should be fine as-is I think, otherwise we can do something like:

```C++
c.src_begin += (uint64_t)((int64_t)src->dim[i].stride * (dst->dim[i].min - src->dim[i].min));
```

The `uint64_t` cast is not strictly needed.

Reply:
Yes, we might want src_begin to decrease in value, and we may want to support negative strides in future. For now we should just write new code assuming they could be negative.

Reply:
OK, but to be clear - `stride` is currently never negative?  I want to make sure this change didn't break anything.

Reply:
Some parts of Halide support negative strides and some don't. device buffer copies currently do not. So in this code, stride is currently not going to be negative (or other things would break too).

Reply:
In theory, negative strides should be fully supported. As you've noticed, though, this is undertested (probably for both normal and `large_buffers` mode). We should add some tests that specifically exercise negative strides to verify that this works (and stays working). (Or declare that we don't support negative strides...)